return warning on empty list result in kubectl get

juanvallejo opened the pull request:
**Release note**:

``` release-note
NONE
```

The current default behavior of `kubectl get` is to return an empty
output when there are no resources to display. This patch improves
usability by returning a warning through stderr in the case of an empty
list.
##### Before

`$ kubectl get pods`
- _empty output_
##### After

`$ kubectl get pods`

```
There are no resources to display.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32722)

<!-- Reviewable:end -->

smarterclayton commented on pkg/kubectl/cmd/get.go:
```diff
@@ -326,6 +326,11 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
+
+		// output empty list warning when no resources to display
+		if len(infos) == 0 {
+			fmt.Fprintf(errOut, "%s %q %s\n", "No resources found in namespace", cmdNamespace, "that match your criteria.")
```
Get can be used on cluster scoped or across multiple namespaces so this message is too specific.  I'd recommend just using a more generic message "No resources found".  The alternative is a more complex integration into resource.Builder - you'd need to accumulate and store markers of intent when the resource builder crafts a Result and then have the Result be able to spit out a non-generic message.  Not sure it's worth it.

fabianofranz replied:
Makes sense, "No resources found" works for me at least for now.

janetkuo commented:
Should we print "No resource found" for the `watch` case also?

I'd suggest making this a function (check `len(infos)==0` and print the warning if so), and put the check right after Infos() is called (fail early).

smarterclayton commented:
With the watch case can the object show up?  I thought we allowed you to
watch something that does not exist yet.

On Thu, Sep 22, 2016 at 6:30 PM, Janet Kuo <notifications@github.com> wrote:

> Should we print "No resource found" for the watch case also?
>
> I'd suggest making this a function (check len(infos)==0 and print the
> warning if so), and put the check right after Infos() is called (fail
> early).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32722#issuecomment-249046981>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p8OJ5kyCvEoeF1ZD3cwsz7sY20Ewks5qswFrgaJpZM4J9Na6>
> .
>

janetkuo commented:
Yeah we allow you to do that. Printing `No resource found` doesn't stop you from keep watching. Or maybe it's not needed for `watch`?

smarterclayton commented:
It might be ok to print "Resource does not yet exist" but that might be
overthinking it.

On Sep 22, 2016, at 6:51 PM, Janet Kuo <notifications@github.com> wrote:

Yeah we allow you to do that. Printing No resource found doesn't stop you
from keep watching. Or maybe it's not needed for watch?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/32722#issuecomment-249051052>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p_wCZf80Bi_Luf3NXXXdUiKxVPeHks5qswZ5gaJpZM4J9Na6>
.

juanvallejo committed:
return warning on empty list result in kubectl get

The current default behavior of `kubectl get` is to return an empty
output when there are no resources to display. This patch improves
usability by returning a warning through stderr in the case of an empty
list.

janetkuo commented on pkg/kubectl/cmd/get.go:
```diff
@@ -327,6 +327,11 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
 
+		// output empty list warning when no resources to display
+		if len(infos) == 0 {
+			fmt.Fprintf(errOut, "%s\n", "No resources found.")
```
Abstract this into a function

janetkuo commented on pkg/kubectl/cmd/get.go:
```diff
@@ -476,5 +481,10 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
+
+	// output empty list warning when no resources to display
+	if len(objs) == 0 {
+		fmt.Fprintf(errOut, "%s\n", "No resources found.")
```
Check for `len(infos)` here, and move both checks right after `infos` is created. 
We can also return right away after no resource found (optional).

juanvallejo commented:
@janetkuo 
> Check for len(infos) here, and move both checks right after infos is created. 
We can also return right away after no resource found (optional).

Thanks for the suggestions! I looked into returning as soon as no resources are found, but this would prevent any additonal errors that may have occurred from printing

janetkuo commented:
LGTM. Please squash the commits and I'll apply the tag.

juanvallejo commented:
@janetkuo 
> LGTM. Please squash the commits and I'll apply the tag.

Done!

juanvallejo commented:
@janetkuo or @smarterclayton PTAL

janetkuo commented:
Need @smarterclayton's approval on changes
